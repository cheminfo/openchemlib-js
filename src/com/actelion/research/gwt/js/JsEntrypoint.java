package com.actelion.research.gwt.js;

import com.google.gwt.core.client.EntryPoint;

public class JsEntrypoint implements EntryPoint {
  @Override
  public void onModuleLoad() {}
}
