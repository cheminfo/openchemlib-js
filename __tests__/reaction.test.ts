import { describe, expect, it } from 'vitest';

import { Molecule, Reaction } from '../lib/index.js';

describe('Reaction class', () => {
  it('should create an empty Reaction', () => {
    const reaction = Reaction.create();
    expect(reaction.getReactants()).toBe(0);
    expect(reaction.getProducts()).toBe(0);
    expect(reaction.getCatalysts()).toBe(0);
  });

  it('should be able to add molecules', () => {
    const reaction = Reaction.create();
    const reactant1 = Molecule.fromSmiles('C');
    const reactant2 = Molecule.fromSmiles('O');
    const product = Molecule.fromSmiles('CO');
    const catalyst = Molecule.fromSmiles('[Fe]');
    reaction.addReactant(reactant1);
    reaction.addReactant(reactant2);
    reaction.addProduct(product);
    reaction.addCatalyst(catalyst);
    expect(reaction.getReactants()).toBe(2);
    expect(reaction.getProducts()).toBe(1);
    expect(reaction.getCatalysts()).toBe(1);
  });

  it('should be able to create from array', () => {
    const reactant1 = Molecule.fromSmiles('C');
    const reactant2 = Molecule.fromSmiles('O');
    const product = Molecule.fromSmiles('CO');
    const reaction = Reaction.fromMolecules([reactant1, reactant2, product], 2);
    expect(reaction.getReactants()).toBe(2);
    expect(reaction.getProducts()).toBe(1);
  });

  it('should work with SMILES', () => {
    const reactant1 = Molecule.fromSmiles('C');
    const reactant2 = Molecule.fromSmiles('O');
    const product = Molecule.fromSmiles('CO');
    const catalyst = Molecule.fromSmiles('[Fe]');
    const reaction = Reaction.fromMolecules([reactant1, reactant2, product], 2);
    reaction.addCatalyst(catalyst);

    const smiles = reaction.toSmiles();
    expect(smiles).toBe('C.O>[Fe]>CO');

    const newReaction = Reaction.fromSmiles(smiles);
    expect(newReaction.getReactants()).toBe(1);
    expect(newReaction.getProducts()).toBe(1);
    expect(newReaction.getCatalysts()).toBe(1);

    expect(newReaction.getReactant(0).toSmiles()).toBe('C.O');
    expect(newReaction.getProduct(0).toSmiles()).toBe('CO');
    expect(newReaction.getCatalyst(0).toSmiles()).toBe('[Fe]');
  });

  it('should load and create RXN files', () => {
    const rxn = `$RXN


JME Molecular Editor
  2  2
$MOL
CC(C)C(=O)Cl.CCNCC>>CCNCC.Cl
JME 2017-11-16 Fri Nov 09 13:48:29 GMT+100 2018

  6  5  0  0  0  0  0  0  0  0999 V2000
    2.4460    0.0000    0.0000 Cl  0  0  0  0  0  0  0  0  0  0  0  0
    2.4350    1.4040    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    3.6307    2.1060    0.0000 O   0  0  0  0  0  0  0  0  0  0  0  0
    1.2175    2.1060    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    1.2066    3.5100    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    0.0000    1.4150    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
  1  2  1  0  0  0  0
  2  3  2  0  0  0  0
  2  4  1  0  0  0  0
  4  5  1  0  0  0  0
  4  6  1  0  0  0  0
M  END
$MOL
CC(C)C(=O)Cl.CCNCC>>CCNCC.Cl
JME 2017-11-16 Fri Nov 09 13:48:29 GMT+100 2018

  5  4  0  0  0  0  0  0  0  0999 V2000
    2.4467    0.0000    0.0000 N   0  0  0  0  0  0  0  0  0  0  0  0
    3.6508    0.7072    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    1.2233    0.7072    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    4.8550    0.0191    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    0.0000    0.0191    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
  1  2  1  0  0  0  0
  1  3  1  0  0  0  0
  2  4  1  0  0  0  0
  3  5  1  0  0  0  0
M  END
$MOL
CC(C)C(=O)Cl.CCNCC>>CCNCC.Cl
JME 2017-11-16 Fri Nov 09 13:48:29 GMT+100 2018

  5  4  0  0  0  0  0  0  0  0999 V2000
    2.4467    0.0000    0.0000 N   0  0  0  0  0  0  0  0  0  0  0  0
    3.6508    0.7072    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    1.2233    0.7072    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    4.8550    0.0191    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    0.0000    0.0191    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
  1  2  1  0  0  0  0
  1  3  1  0  0  0  0
  2  4  1  0  0  0  0
  3  5  1  0  0  0  0
M  END
$MOL
CC(C)C(=O)Cl.CCNCC>>CCNCC.Cl
JME 2017-11-16 Fri Nov 09 13:48:29 GMT+100 2018

  1  0  0  0  0  0  0  0  0  0999 V2000
    0.0000    0.0000    0.0000 Cl  0  0  0  0  0  0  0  0  0  0  0  0
M  END
`;

    const reaction = Reaction.fromRxn(rxn);
    expect(reaction.getReactants()).toBe(2);
    expect(reaction.getProducts()).toBe(2);
    expect(reaction.getCatalysts()).toBe(0);

    expect(reaction.toRxn()).toMatchSnapshot();

    expect(reaction.toRxnV3()).toMatchSnapshot();
  });
});
