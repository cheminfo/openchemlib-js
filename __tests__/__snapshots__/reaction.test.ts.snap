// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Reaction class > should load and create RXN files 1`] = `
"$RXN


OCL_RXN_V1.0:gGQHDHaInfh@ gJX@@eKU@@!gJX@@eKU@@ fHdP@@##!R_\`CW?Hc|_\`BH_Ykh@@ !ROvp@[C}c@FNZz@@ !ROvp@[C}c@FNZz@@ !RGP@@##
  2  2
$MOL
CC(C)C(=O)Cl.CCNCC>>CCNCC.Cl
Actelion Java MolfileCreator 1.0

  6  5  0  0  0  0  0  0  0  0999 V2000
    1.7471    0.0000    0.0000 Cl  0  0  0  0  0  0  0  0  0  0  0  0
    1.7393    1.0029    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    2.5934    1.5043    0.0000 O   0  0  0  0  0  0  0  0  0  0  0  0
    0.8696    1.5043    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    0.8619    2.5072    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    0.0000    1.0107    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
  1  2  1  0  0  0  0
  2  3  2  0  0  0  0
  2  4  1  0  0  0  0
  4  5  1  0  0  0  0
  4  6  1  0  0  0  0
M  END
$MOL
CC(C)C(=O)Cl.CCNCC>>CCNCC.Cl
Actelion Java MolfileCreator 1.0

  5  4  0  0  0  0  0  0  0  0999 V2000
    1.7476    0.0000    0.0000 N   0  0  0  0  0  0  0  0  0  0  0  0
    2.6077    0.5051    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    0.8738    0.5051    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    3.4679    0.0136    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    0.0000    0.0136    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
  1  2  1  0  0  0  0
  1  3  1  0  0  0  0
  2  4  1  0  0  0  0
  3  5  1  0  0  0  0
M  END
$MOL
CC(C)C(=O)Cl.CCNCC>>CCNCC.Cl
Actelion Java MolfileCreator 1.0

  5  4  0  0  0  0  0  0  0  0999 V2000
    1.7476    0.0000    0.0000 N   0  0  0  0  0  0  0  0  0  0  0  0
    2.6077    0.5051    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    0.8738    0.5051    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    3.4679    0.0136    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    0.0000    0.0136    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
  1  2  1  0  0  0  0
  1  3  1  0  0  0  0
  2  4  1  0  0  0  0
  3  5  1  0  0  0  0
M  END
$MOL
CC(C)C(=O)Cl.CCNCC>>CCNCC.Cl
Actelion Java MolfileCreator 1.0

  1  0  0  0  0  0  0  0  0  0999 V2000
    0.0000    0.0000    0.0000 Cl  0  0  0  0  0  0  0  0  0  0  0  0
M  END
"
`;

exports[`Reaction class > should load and create RXN files 2`] = `
"$RXN V3000


OCL_RXN_V1.0:gJX@@eKU@@ gGQHDHaInfh@!gJX@@eKU@@ fHdP@@##!ROvp@[C}c@FNZz@@ !R_\`CW?Hc|_\`BH_Ykh@@ !ROvp@[C}c@FNZz@@ !RGP@@##
M  V30 COUNTS 2 2
M  V30 BEGIN REACTANT
M  V30 BEGIN CTAB
M  V30 COUNTS 6 5 0 0 0
M  V30 BEGIN ATOM
M  V30 1 Cl 1.7471 0 0 0
M  V30 2 C 1.7392 1.0028 0 0
M  V30 3 O 2.5933 1.5042 0 0
M  V30 4 C 0.8696 1.5042 0 0
M  V30 5 C 0.8618 2.5071 0 0
M  V30 6 C 0 1.0107 0 0
M  V30 END ATOM
M  V30 BEGIN BOND
M  V30 1 1 1 2
M  V30 2 2 2 3
M  V30 3 1 2 4
M  V30 4 1 4 5
M  V30 5 1 4 6
M  V30 END BOND
M  V30 END CTAB
M  V30 BEGIN CTAB
M  V30 COUNTS 5 4 0 0 0
M  V30 BEGIN ATOM
M  V30 1 N 1.7476 0 0 0
M  V30 2 C 2.6077 0.5051 0 0
M  V30 3 C 0.8737 0.5051 0 0
M  V30 4 C 3.4678 0.0136 0 0
M  V30 5 C 0 0.0136 0 0
M  V30 END ATOM
M  V30 BEGIN BOND
M  V30 1 1 1 2
M  V30 2 1 1 3
M  V30 3 1 2 4
M  V30 4 1 3 5
M  V30 END BOND
M  V30 END CTAB
M  V30 END REACTANT
M  V30 BEGIN PRODUCT
M  V30 BEGIN CTAB
M  V30 COUNTS 5 4 0 0 0
M  V30 BEGIN ATOM
M  V30 1 N 1.7476 0 0 0
M  V30 2 C 2.6077 0.5051 0 0
M  V30 3 C 0.8737 0.5051 0 0
M  V30 4 C 3.4678 0.0136 0 0
M  V30 5 C 0 0.0136 0 0
M  V30 END ATOM
M  V30 BEGIN BOND
M  V30 1 1 1 2
M  V30 2 1 1 3
M  V30 3 1 2 4
M  V30 4 1 3 5
M  V30 END BOND
M  V30 END CTAB
M  V30 BEGIN CTAB
M  V30 COUNTS 1 0 0 0 0
M  V30 BEGIN ATOM
M  V30 1 Cl 0 0 0 0
M  V30 END ATOM
M  V30 BEGIN BOND
M  V30 END BOND
M  V30 END CTAB
M  V30 END PRODUCT
M  END
"
`;
